Php config imports

loic425 opened the pull request:
Referenced issue : #1539 
Based on #1537 

Several proposals for this importing feature.

**Use variadic**
```php
use Behat\Config\Config;

$config = new Config();
$config->import('first_file.php', 'second_file.php', ...);
```

If we have too much files to imports, using multiple lines is still possible.

```php
use Behat\Config\Config;

$config = new Config();
$config->import(
    'first_file.php', 
    'second_file.php', 
    ...,
    ...,
    'another_file.php', 
);
```

**Use array**
```php
use Behat\Config\Config;

$config = new Config();
$config->import(['first_file.php', 'second_file.php', ...]);
```

**[Current state of this PR] Use simple way (and even support array or string at the same time)**
```php
use Behat\Config\Config;

$config = new Config();
$config
    ->import('first_file.php') 
    ->import('second_file.php')
    // ...
;
```

OR

```php
use Behat\Config\Config;

$config = new Config();
$config->import(['first_file.php', 'second_file.php', ...]) ;
```

One of the benefits of this way is that we can add optional arguments later.

```php
public function import(string|array $resource, string $type = null, bool $ignoreErrors = false): self
```

Moreover, we need to think about an implementation for importing using a pattern. 

**Using pattern**
```php
use Behat\Config\Config;

$config = new Config();
$config->import('config/../..php');
```

**Symfony way**

_Dependency Injection_
https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php#L64

_Routing_
The routing way is also interesting
https://github.com/symfony/routing/blob/7.1/Loader/Configurator/RoutingConfigurator.php#L37

loic425 commented on src/Behat/Config/Config.php:
```diff
@@ -0,0 +1,25 @@
+    ) {
+    }
+
+    public function import(string $file): self
```
Could be interesting...

```suggestion
    public function import(... string $files): self
```

acoulton commented on src/Behat/Config/Config.php:
```diff
@@ -0,0 +1,27 @@
+    public function import(string ...$files): self
+    {
+        foreach ($files as $file) {
+            $this->settings['imports'][] = $file;
```
I might have missed something, but if the `->import()` method just adds keys/values to the array you are already defining in the file, why wouldn't you just do that inline?

E.g. 
```
return new Config([
  'imports' => [
    __DIR__.'/other-config.php'
  ],
  // Other values 
])
```

acoulton reviewed (changes requested):
Obviously we need to look at #1537 first, but probably worth considering the overall feature / plan for future direction

loic425 commented on src/Behat/Config/Config.php:
```diff
@@ -0,0 +1,27 @@
+    ) {
+    }
+
+    public function import(string ...$files): self
```
I think this should be better without variadic thing or I should remame the method "imports" instead.
```suggestion
    public function import(string $file): self
```

acoulton commented:
I think my preference would be:
* For explicit files, a variadic e.g. `import(string ...$imports)`
* For pattern matching, a separate `importMatching(string $pattern)` - if you want multiple patterns, you call it multiple times. Possibly with something like `importMatching(string $pattern, bool $throwIfNoMatches = true)` to give people control / safety over what happens if nothing matches the pattern.

loic425 commented:
> I think my preference would be:
> 
>     * For explicit files, a variadic e.g. `import(string ...$imports)`
> 
>     * For pattern matching, a separate `importMatching(string $pattern)` - if you want multiple patterns, you call it multiple times. Possibly with something like `importMatching(string $pattern, bool $throwIfNoMatches = true)` to give people control / safety over what happens if nothing matches the pattern.

I've removed the two methods way from the proposal and added a simple one based on Symfony routing configurator.

loic425 commented on src/Behat/Config/Config.php:
```diff
@@ -4,13 +4,26 @@
     ) {
     }
 
+    public function import(string|array $resource): self
```
```suggestion
    /** @param string|string[] $resource **/
    public function import(string|array $resource): self
```

carlos-granados reviewed (approved):
Looking good to me, thanks @loic425

loic425 commented:
@acoulton Is the current state ok for you.
I've simplified with a single method with string or array of string.

acoulton commented on src/Behat/Config/Config.php:
```diff
@@ -4,13 +4,27 @@
     ) {
     }
 
+    /** @param string|string[] $resource **/
+    public function import(string|array $resource): self
+    {
+        $resources = is_string($resource) ? [$resource] : $resource;
```
I would personally use a variadic here, it gives us (and users) strict typing out of the box including protecting against unexpected/unused extra arguments, keeps our own code as thin as possible, and therefore reduces any risk of edge cases.

As a user, the syntax `->import('one.php', 'two.php')` is also cleaner and easier to format IMO than having the redundant array braces.

I know there are other tools out there that do `string|string[]` but my sense is those are interfaces that were designed before variadics were available and so they're tied to them. I don't think we should be, unless there's a reason not to use a variadic that I haven't thought of?

acoulton replied:
Fair enough, makes sense.
